### PR TITLE
Add tpmdata file io functions for keylime to read/write tpm data

### PR DESCRIPTION
**Pull request by leonjia0112**
Added four functions for read/white/get/set tpmdata.json file where tpm data is stored in tpm.rs. Also added a tpmdata_test.json for testing purpose.

* Changed .travis.yml to single thread testing, because parallel testing causes the file changed by other tests and fail.
* Modified tpmdata read function with error handling without panic.

Compare to Lincoln lab version:
Original python version used a global variable tpmdata to store tpmdata in global scope. This implementation eliminated global variable is to store tpm data since it is not a frequent access variable.

## Test passed. Ready for a review. Thanks.

### CODE REVIEW

**Comment by leonjia0112** [comment]
Test passed. Ready for another review. I changed Travis back to what it was, and now all the tests should pass without fail. Thanks.

File: src/tpm.rs [hunk 1]
```diff
@@ -206,9 +272,9 @@ pub fn create_deep_quote(
     mut vpcrmask: String,
 ) -> Option<String> {
     let quote_path = NamedTempFile::new().unwrap();
-    let key_handle = get_tpm_metadata("aik_handle".to_string());
-    let aik_password = get_tpm_metadata("aik_pw".to_string());
-    let owner_password = get_tpm_metadata("owner_pw".to_string());
+    let key_handle = get_tpm_metadata_content("aik_handle").unwrap();
```
Comment (frozencemetery) [inline src/tpm.rs]: Do not call `unwrap` or `panic!` outside the test suite.

File: src/tpm.rs [hunk 2]
```diff
@@ -26,50 +28,114 @@ pub const EXIT_SUCCESS: i32 = 0;
...
-    let t_data = read_tpm_data().unwrap();
-    let t_data_json_value = serde_json::to_value(t_data).unwrap();
-    Some(t_data_json_value[content_key].to_string())
+ * Input: content key in tpmdata
+ * Return: Result wrap value string or error code -1
+ *
+ * Getting the tpm data struct and convert it to a json value object to
+ * retrive a particular value by the given key inside the tpm data.
+ */
+fn get_tpm_metadata_content(key: &str) -> Result<Option<String>, i32> {
```
Comment (frozencemetery) [inline src/tpm.rs]: This type is kind of gross.  I think you can do it with a `Result<String, i32>` (see below).  However, the i32 doesn't seem to be meaningful; all you ever put there is -1.

File: src/tpm.rs [hunk 3]
```diff
@@ -26,50 +28,114 @@ pub const EXIT_SUCCESS: i32 = 0;
...
+            content
+                .as_str()
+                .unwrap()
+                .to_string()
+                .trim_matches(remove)
+                .to_string(),
+        )),
+        None => {
+            debug!("Content field [{}] doesn't exist.", key);
+            Ok(None)
```
Comment (frozencemetery) [inline src/tpm.rs]: Should be an `Err`, yeah?

File: src/tpm.rs [hunk 4]
```diff
@@ -525,6 +591,54 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_get_tpm_metadata_1() {
+        match set_tpmdata_test() {
+            Ok(_) => {
+                // using test tpmdata.json content must present, system won't panic
```
Comment (frozencemetery) [inline src/tpm.rs]: Instead of panicing, consider printing a useful error message.

File: src/tpm.rs [hunk 5]
```diff
@@ -525,6 +591,54 @@ mod tests {
...
+                    // using test tpmdata.json content must present, system won't panic
+                    let remove: &[_] = &['"', ' ', '/'];
+                    let password =
+                        get_tpm_metadata_content("owner_pw").unwrap();
+                    assert_eq!(
+                        password.unwrap().trim_matches(remove),
+                        String::from("hello")
+                    )
+                }
+                Err(_) => assert!(false),
```
Comment (frozencemetery) [inline src/tpm.rs]: Instead of this, I would do something like: `assert!(set_tpmdata_test().is_ok())`

File: src/tpm.rs (hunk 1 above)
Comment (leonjia0112) [reply src/tpm.rs]: Handled error and return None if an error occurs.

File: src/tpm.rs (hunk 2 above)
Comment (leonjia0112) [reply src/tpm.rs]: Changed to Result<String, String>

File: src/tpm.rs (hunk 3 above)
Comment (leonjia0112) [reply src/tpm.rs]: Changed to Err for a little simple implementation.

File: src/tpm.rs (hunk 4 above)
Comment (leonjia0112) [reply src/tpm.rs]: Print out an error message explaining what happened.

File: src/tpm.rs (hunk 5 above)
Comment (leonjia0112) [reply src/tpm.rs]: This is great.  Changed.

**Comment by leonjia0112** [comment]
Changed based on review:
1. Changed error from -1 to message that explains the detail of the error.
2. Changed Option<String> to String
3. Used is_ok() in testing
4. Remove .unwrap() that panic the system with error handling and early return.

**Test Passed. Ready for a review.**
Thanks.

**Comment by leonjia0112** [comment]
Now there are only tpm.rs and tpmdata_test.json files are changed in this PR.

Note: 
tpmdata struct was removed because it is no longer used by the read_tpm_data() function.
File changed of main.rs and common.rs is because travis is complaining about the formatting, which it wasn't complaining before.

**Test passed, ready for a review**
Thanks

File: src/tpm.rs [hunk 6]
```diff
@@ -26,50 +27,160 @@ pub const EXIT_SUCCESS: i32 = 0;
...
+            error!("Fail to read tpmdata.json.");
+            return Err(e);
+        }
+    };
+
+    let remove: &[_] = &['"', ' ', '/'];
+    match tpm_data.get(key) {
+        Some(content) => Ok(content
+            .as_str()
+            .unwrap()
```
Comment (frozencemetery) [inline src/tpm.rs]: unwrap is not permitted here

File: src/tpm.rs [hunk 7]
```diff
@@ -26,50 +27,160 @@ pub const EXIT_SUCCESS: i32 = 0;
...
+ * caller for error handling.
  */
-fn read_tpm_data() -> Result<TpmData, Box<Error>> {
-    let file = File::open("tpmdata.json")?;
-    let data: TpmData = serde_json::from_reader(file)?;
+fn read_tpm_data() -> Result<Value, Box<String>> {
+    let file = match File::open("tpmdata.json") {
+        Ok(f) => f,
+        Err(e) => {
+            return Err(Box::new(format!(
```
Comment (frozencemetery) [inline src/tpm.rs]: Can you make a helper function for this `Err(Box::new(format!()))` gunk?

**Comment by leonjia0112** [comment]
Add a output_error_message() function for Err(Box::new(foramt!(...))) that can take a error message and a error and return a error message string. 

.unwrap()s are removed with error handling Result<>.

** Test passed. Ready for a review.**
Thanks!

**Comment by frozencemetery** [comment]
@leonjia0112 I've made some changes.  Please take a look and if you're satisfied they're okay, I'll merge.

**Comment by leonjia0112** [comment]
Those changes are decent. I am ok with those changes. Thanks :)

**Pull request merged**